code_samples/project_view_pane/src/main/java/org/intellij/sdk/view/pane/ImagesProjectNode.java

```diff
@@ -75,15 +79,51 @@ private Set<VirtualFile> getImagesFiles(Project project) {
     return files;
   }
 
+  private void setupImageFilesRefresher(@NotNull Project project, @NotNull Disposable parentDisposable) {
+    project.getMessageBus().connect(parentDisposable)
+        .subscribe(VirtualFileManager.VFS_CHANGES, new BulkFileListener() {
+          @Override
+          public void after(@NotNull List<? extends @NotNull VFileEvent> events) {
+            boolean hasAnyImageUpdate = false;
+            for (VFileEvent event : events) {
+              VirtualFile file = event.getFile();
+              if (file == null) {
+                continue;
+              }
+              String extension = file.getExtension();
+              if (extension != null && SUPPORTED_IMAGE_EXTENSIONS.contains(extension)) {
+                hasAnyImageUpdate = true;
+                break;
+              }
+            }
+            if (hasAnyImageUpdate) {
+              getImagesFiles(project).clear();
+              scanImages(project);
+              ApplicationManager.getApplication().invokeLater(() ->
```

Comment:
use `com.intellij.util.ui.update.MergingUpdateQueue` to prevent too many/"overlapping" updates in UI

Reply:
use `invokeLater` with `Project.getDisposed` condition overload

Reply:
> use `com.intellij.util.ui.update.MergingUpdateQueue` to prevent too many/"overlapping" updates in UI

I think it should not be a problem, as we use `BulkFileListener`, and UI update happens once for bulk if there is at least one image change. Or am I missing something?

Reply:
You were right about using `MergingUpdateQueue`. Actually, even though I used `BulkFileListener`, bulks were unexpectedly for me created for each file and UI was updated multiple times. Thanks.

The rest of the issues have been also fixed.